### PR TITLE
Highmaps

### PR DESCRIPTION
**Pull request by LinghanX**

### CODE REVIEW

File: app/controllers/bootcamps_controller.rb [hunk 1]
```diff
@@ -16,7 +16,72 @@ def index
...
+            ['TN'],
+            ['TX'],
+            ['UT'],
+            ['VT'],
+            ['VA'],
+            ['WA'],
+            ['WV'],
+            ['WI'],
+            ['WY']
+          ]
```
Comment (radavis) [inline app/controllers/bootcamps_controller.rb]: I would move the states helper to `lib`. 

Add this line to `application.rb`:

```
config.autoload_paths << Rails.root.join('lib')
```
Make the us_states method a class method (add `self.`).

Now, you can:

```
us_state_abbreviations = StatesHelper::us_states.map { |state| [state[1]] }
```

File: app/helpers/states_helper.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
+module StatesHelper
+  def us_states
```
Comment (radavis) [inline app/helpers/states_helper.rb]: self.us_states

File: app/views/bootcamps/_form.html.erb [hunk 3]
```diff
@@ -24,6 +24,9 @@
   <%= f.label :duration %>
   <%= f.number_field :duration, placeholder: 'please input a number, say 10 for 10 weeks.' %>
 
+  <%= f.label :state %>
+  <%= f.select :state, options_for_select(us_states) %>
```
Comment (radavis) [inline app/views/bootcamps/_form.html.erb]: StatesHelper::us_states

File: db/seeds.rb [hunk 4]
```diff
@@ -23,15 +11,72 @@
...
+    ['WY']
+  ]
+
   Bootcamp.create!(name: Faker::Name.name, city: Faker::Address.city,
                    country: "USA", duration: 3, price: rand(4000..18000),
                    description: Faker::Hacker.say_something_smart,
-                   rating: rand(11), is_online: true, user_id: rand(21))
+                   rating: rand(11), is_online: true,
+                   user_id: rand(21),
+                   state: us_states[rand(us_states.size)][0])
```
Comment (radavis) [inline db/seeds.rb]: `States::us_states.sample[1]`

File: db/seeds.rb [hunk 5]
```diff
@@ -23,15 +11,72 @@
...
 end
 
-100.times do
+2000.times do
   Review.create(title: Faker::Lorem.sentence, body: Faker::Lorem.paragraph,
-                rating: rand(11), bootcamp_id: rand(10) + 1,
+                rating: rand(11), bootcamp_id: rand(199) + 1,
```
Comment (radavis) [inline db/seeds.rb]: `Bootcamp.all.sample`

File: app/controllers/bootcamps_controller.rb [hunk 6]
```diff
@@ -16,7 +16,72 @@ def index
           bootcamp_prices = @bootcamps.map { |bootcamp| bootcamp.price.to_i }
           bootcamp_ratings = @bootcamps.map { |bootcamp| bootcamp.ave_rating }
 
-          render json: [bootcamp_names, bootcamp_prices, bootcamp_ratings]
+          bootcamp_map_date = Hash.new
+
+          us_states = [
```
Comment (dreamingfurther) [inline app/controllers/bootcamps_controller.rb]: this needs to be extracted out of the controller. A good pattern for values like this is a constant in the model So in the controller you'd do something like:

```
us_states = Bootcamp::US_STATES
```

and then have the array of arrays in your model.

File: app/controllers/bootcamps_controller.rb (hunk 1 above)
Comment (dreamingfurther) [inline app/controllers/bootcamps_controller.rb]: Richard's suggestion is also valid. Pick one way to refactor this

File: db/seeds.rb [hunk 7]
```diff
@@ -23,15 +11,72 @@
               password: Faker::Internet.password(8))
 end
 
-10.times do
+200.times do
```
Comment (dreamingfurther) [inline db/seeds.rb]: why are you seeding 10,000 states?

**Pull request merged**